src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -1106,6 +1115,116 @@ CreateSchemaDDLCommand(Oid schemaId)
 }
 
 
+/*
+ * GrantOnSchemaDDLCommands creates a list of ddl command for replicating the permissions
+ * of roles on schemas.
+ */
+List *
+GrantOnSchemaDDLCommands(Oid schemaOid)
+{
+	HeapTuple schemaTuple = SearchSysCache1(NAMESPACEOID, ObjectIdGetDatum(schemaOid));
+	bool isNull = true;
+	Datum aclDatum = SysCacheGetAttr(NAMESPACEOID, schemaTuple, Anum_pg_namespace_nspacl,
+									 &isNull);
+	if (isNull)
+	{
+		ReleaseSysCache(schemaTuple);
+		return NIL;
+	}
+	Acl *acl = DatumGetAclPCopy(aclDatum);
+	AclItem *aclDat = ACL_DAT(acl);
+	int aclNum = ACL_NUM(acl);
+	List *commands = NIL;
+	Oid currentUser = GetUserId();
+
+	ReleaseSysCache(schemaTuple);
+
+	for (int i = 0; i < aclNum; i++)
+	{
+		bool withSetRole = (currentUser != aclDat[i].ai_grantor);
+		commands = list_concat(commands,
+							   GenerateGrantOnSchemaQueriesFromAclItem(
+								   schemaOid,
+								   &aclDat[i],
+								   withSetRole));
+	}
+
+	return commands;
+}
+
+
+/*
+ * GenerateGrantOnSchemaQueryFromACL generates a query string for replicating a users permissions
+ * on a schema.
+ */
+List *
+GenerateGrantOnSchemaQueriesFromAclItem(Oid schemaOid, AclItem *aclItem, bool withSetRole)
+{
+	AclMode permissions = ACLITEM_GET_PRIVS(*aclItem) & ACL_ALL_RIGHTS_SCHEMA;
+	AclMode grants = ACLITEM_GET_GOPTIONS(*aclItem) & ACL_ALL_RIGHTS_SCHEMA;
+	Assert(!(grants & ACL_USAGE) || (permissions & ACL_USAGE));
```

Comment:
can you leave a comment here what the asserts are asserting for?

Reply:
Since this is asserting on two different variables, better to make this 2 asserts

Reply:
But there is an || in between. I need at least one of them to be true, not both.

Reply:
Right, my bad